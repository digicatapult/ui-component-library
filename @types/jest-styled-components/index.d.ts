declare module 'jest-styled-components/serializer'
