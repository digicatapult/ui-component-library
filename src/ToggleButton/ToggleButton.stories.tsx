import React, { useState } from 'react'
import { Story } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import ToggleButton from './index.js'

export default {
  title: 'Components/ToggleButton',
  component: ToggleButton,
  argTypes: {
    height: {
      control: 'text',
    },
    width: {
      control: 'text',
    },
    borderRadius: {
      control: 'text',
    },
    isOpen: {
      control: 'boolean',
    },
    openBackground: {
      control: 'color',
    },
    closeBackground: {
      control: 'color',
    },
    openContent: {
      control: 'text',
    },
    closeContent: {
      control: 'text',
    },
    fontColor: {
      control: 'color',
    },
    fontWeight: {
      control: 'text',
    },
  },
}

const DefaultStoryTemplate: Story = (args) => {
  const [isOpen, setIsOpen] = useState(false)

  const onClick = () => setIsOpen(!isOpen)

  return <ToggleButton onClick={onClick} isOpen={isOpen} {...args} />
}

export const Default = DefaultStoryTemplate.bind({})
Default.args = {
  height: '40px',
  width: '80px',
  closeBackground: 'lightyellow',
  borderRadius: '1em',
  closeContent: 'Help',
  openContent: 'Close',
}
export const HII = DefaultStoryTemplate.bind({})
HII.args = {
  height: '2.5em',
  width: '2.5em',
  closeBackground: '#27847A',
  openBackground: '#000',
  borderRadius: '1.25em',
  fontColor: '#FFF',
  fontWeight: 'bold',
}
