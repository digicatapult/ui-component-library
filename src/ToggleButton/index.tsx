import React from 'react'
import styled from 'styled-components'

export interface ToggleButtonProps {
  width?: string
  height?: string
  borderRadius?: string
  isOpen?: boolean
  openBackground?: string
  closeBackground?: string
  openContent?: string
  closeContent?: string
  fontColor?: string
  fontWeight?: string
  onClick?: () => void
}

const ToggleButton: React.FC<ToggleButtonProps> = ({
  width = 'auto',
  height = 'auto',
  borderRadius = '',
  isOpen = false,
  openBackground = 'lightgray',
  closeBackground = '#FFF',
  openContent = 'X',
  closeContent = '?',
  fontColor = 'black',
  fontWeight = 'inherit',
  onClick,
}) => {
  return (
    <Wrapper
      width={width}
      height={height}
      borderRadius={borderRadius}
      isOpen={isOpen}
      openBackground={openBackground}
      closeBackground={closeBackground}
      openContent={openContent}
      closeContent={closeContent}
      fontColor={fontColor}
      fontWeight={fontWeight}
      onClick={onClick}
    >
      {isOpen ? openContent : closeContent}
    </Wrapper>
  )
}

const Wrapper = styled.button<ToggleButtonProps>`
  width: ${({ width }) => width};
  height: ${({ height }) => height};
  border: 0;
  cursor: pointer;
  color: ${({ fontColor }) => fontColor};
  font-weight: ${({ fontWeight }) => fontWeight};

  background: ${({ isOpen, openBackground, closeBackground }) =>
    isOpen ? openBackground : closeBackground};
  border-radius: ${({ borderRadius }) => borderRadius};
`

export default ToggleButton
