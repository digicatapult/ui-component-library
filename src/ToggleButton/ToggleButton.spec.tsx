import React, { useState } from 'react'
import renderer, { act } from 'react-test-renderer'
import 'jest-styled-components'

import ToggleButton from './index.js'

describe('Test', () => {
  test('default', () => {
    const clickHandler = jest.fn(() => {})
    const tree = renderer
      .create(<ToggleButton onClick={clickHandler} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('hii', () => {
    const clickHandler = jest.fn(() => {})
    const tree = renderer
      .create(
        <ToggleButton
          height="2.5em"
          width="2.5em"
          closeBackground="#27847A"
          openBackground="#000"
          borderRadius="1.25em"
          fontColor="#FFF"
          fontWeight="bold"
          onClick={clickHandler}
        />,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('click', async () => {
    const openContent = 'opened'
    const closeContent = 'closed'

    const Wrapper = () => {
      const [isOpen, setIsOpen] = useState(false)
      const onClick = () => setIsOpen(!isOpen)
      return (
        <ToggleButton
          openContent={openContent}
          closeContent={closeContent}
          isOpen={isOpen}
          onClick={onClick}
        />
      )
    }

    const instance = renderer.create(<Wrapper />)

    await act(() => {
      instance.root.findByType('button').props.onClick({})
    })

    expect(instance.root.findByType('button').children).toEqual([openContent])

    await act(() => {
      instance.root.findByType('button').props.onClick({})
    })

    expect(instance.root.findByType('button').children).toEqual([closeContent])
  })
})
