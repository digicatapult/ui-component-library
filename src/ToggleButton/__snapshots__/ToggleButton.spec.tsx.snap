// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test default 1`] = `
.c0 {
  width: auto;
  height: auto;
  border: 0;
  cursor: pointer;
  color: black;
  font-weight: inherit;
  background: #FFF;
}

<button
  borderRadius=""
  className="c0"
  closeBackground="#FFF"
  closeContent="?"
  fontColor="black"
  fontWeight="inherit"
  height="auto"
  isOpen={false}
  onClick={[MockFunction]}
  openBackground="lightgray"
  openContent="X"
  width="auto"
>
  ?
</button>
`;

exports[`Test hii 1`] = `
.c0 {
  width: 2.5em;
  height: 2.5em;
  border: 0;
  cursor: pointer;
  color: #FFF;
  font-weight: bold;
  background: #27847A;
  border-radius: 1.25em;
}

<button
  borderRadius="1.25em"
  className="c0"
  closeBackground="#27847A"
  closeContent="?"
  fontColor="#FFF"
  fontWeight="bold"
  height="2.5em"
  isOpen={false}
  onClick={[MockFunction]}
  openBackground="#000"
  openContent="X"
  width="2.5em"
>
  ?
</button>
`;
